### PR TITLE
address a nasty protected[pkg] class regression

### PR DESCRIPTION
**Pull request by retronym**
Shows up in combination with default args.

Review by @paulp for the code, and by @adriaanm for my wildly optimistic target branch.

### CODE REVIEW

**Comment by adriaanm** [comment]
Do we know whether the reporter was able to work around the problem by dropping either component of this lethal combo?

**Comment by retronym** [comment]
They had to make the classes public as a workaround.

**Comment by holograph** [comment]
Aye, that we did. I'd be happy to test a potential patch - Jason, is the second commit (37917d7) supposed to fix the problem? I can test it very quickly if there's any chance of it making its way into 2.10.1.

File: src/compiler/scala/tools/nsc/typechecker/SuperAccessors.scala [hunk 1]
```diff
@@ -288,6 +291,7 @@ abstract class SuperAccessors extends transform.Transform with transform.TypingT
                      currentClass.isTrait
                   && sym.isProtected
                   && sym.enclClass != currentClass
+                  && !sym.owner.isPackageClass
                   && !sym.owner.isTrait
```
Comment (paulp) [inline src/compiler/scala/tools/nsc/typechecker/SuperAccessors.scala:295]:
```
// FIXME - this should be unified with needsProtectedAccessor, but some
// subtlety which presently eludes me is foiling my attempts.
...
[needsProtectedAccessor]
isCandidate && !host.isPackageClass && !isSelfType && !isJavaProtected
```

Sigh, code duplication FTL once again.

**Comment by paulp** [comment]
Looks good to me, speaking target-branch-agnostically.

**Comment by adriaanm** [comment]
it would be good to document the reasoning behind this a bit more
what does the `sym.owner.isPackageClass` tell us exactly?

why look at `qual.symbol` one time and `tree.symbol.owner` the other (for `tree == Select(qual, _)`)

**Comment by adriaanm** [comment]
I thought about it a bit more and since this is a regression from before 2.10.0, we shouldn't fix it here at this late hour. Sorry. It'll be available in a 2.10.x nightly soon.

**Comment by retronym** [comment]
``` scala
package p
protected[p] class C(a: Any = 0)
protected[p] object O {
  new C() // expands approx. to { val arg0 = Select(Select(This(p), newTermName("C"))), newTermName("default$0")); new C(arg0)
}
```

In this patch, we detect if `p2.this.C` is top level , which means that it's qualifier's symbol, package `p2`, is not a suitable or necessary host for a super accessor. `qualifier.symbol.isPackageClass` would tell is the same thing in this case. Is that always the case? I wasn't sure, so I opted to fit in with the existing checks.

**Comment by retronym** [comment]
btw, were we on master, we would have written the condition as: `!sym.isTopLevel`

**Pull request closed without merging**